checker/check.go

```diff
@@ -0,0 +1,145 @@
+
+type Action int
+
+const (
+	ALLOW Action = iota
+	DENY
+	LOG
+	PASS
+	NO_MATCH // Indicates policy did not match request. Cannot be assigned to rule.
+)
+
+// Check a list of policies and return OK if the check passes, or PERMISSION_DENIED if the check fails.
+// Note, if no policy matches, the default is PERMISSION_DENIED.
+func checkStore(store *policystore.PolicyStore, req *authz.CheckRequest) (s status.Status) {
+	s = status.Status{Code: PERMISSION_DENIED}
+	ep := store.Endpoint
+	if ep == nil {
+		log.Warning("CheckRequest before we synced Endpoint information.")
+		return
+	}
+	if len(ep.Tiers) > 0 {
+		log.Debug("Checking policy tier 1.") // We only support a single tier.
+
+		tier := ep.Tiers[0]
+		policies := tier.IngressPolicies
+	Policy:
+		for i, name := range policies {
+			pID := proto.PolicyID{Tier: tier.GetName(), Name: name}
+			policy := store.PolicyByID[pID]
+			action := checkPolicy(policy, req)
+			log.Debugf("Policy %d %v returned action %v", i, pID, action)
+			switch action {
+			case NO_MATCH:
+				continue
+			// If the Policy matches, end evaluation (skipping profiles, if any)
+			case ALLOW:
+				s.Code = OK
+				return
+			case DENY:
+				s.Code = PERMISSION_DENIED
+				return
+			case PASS:
+				// Pass means end evaluation of policies and proceed to profiles, if any.
+				break Policy
+			case LOG:
+				panic("policy should never return LOG action")
+			}
+		}
+	}
+	// If we reach here, there were either no policies, or none that matched.
+	if len(ep.ProfileIds) > 0 {
+		for i, name := range ep.ProfileIds {
+			pID := proto.ProfileID{Name: name}
+			profile := store.ProfileByID[pID]
+			action := checkProfile(profile, req)
+			log.Debugf("Profile %d %v returned action %v", i, name, action)
+			switch action {
+			case NO_MATCH:
+				continue
+			case ALLOW:
+				s.Code = OK
+				return
+			case DENY, PASS:
+				s.Code = PERMISSION_DENIED
+				return
+			case LOG:
+				panic("profile should never return LOG action")
+			}
+		}
+	} else {
+		log.Debug("0 active profiles, deny request.")
+		s.Code = PERMISSION_DENIED
+	}
+	return
+}
+
+// checkPolicy checks if the policy matches the request data, and returns the action.
+func checkPolicy(policy *proto.Policy, req *authz.CheckRequest) (action Action) {
+	// Note that we support only inbound policy.
+	return checkRules(policy.InboundRules, req)
+}
+
+func checkProfile(p *proto.Profile, req *authz.CheckRequest) (action Action) {
+	return checkRules(p.InboundRules, req)
+}
+
+func checkRules(rules []*proto.Rule, req *authz.CheckRequest) (action Action) {
+	for _, r := range rules {
+		if match(r, req) {
+			log.Debugf("Rule matched.")
+			a := ActionFromString(r.Action)
+			if a != LOG {
+				// We don't support actually logging requests, but if we hit a LOG action, we should
+				// continue processing rules.
+				return a
+			}
+		}
+	}
+	return NO_MATCH
+}
+
+func ActionFromString(s string) Action {
+	m := map[string]Action{
```

Comment:
Same here, if you use libcalico-go API action definition, it'd be easier and won't have to be maintained at two different places: https://github.com/projectcalico/libcalico-go/blob/master/lib/apis/v3/policy.go#L160-L165